docs/design/coreclr/jit/Jit Stack Security.md

```diff
@@ -0,0 +1,64 @@
+# .NET Stack Security
+
+This document describes mechanisms in the .NET code generator to ensure stack security at runtime.
+
+## Background
+
+While .NET is primarily a type and memory safe "managed" programming platform, it also offers lower-level
+facilities to allow for interop with native code, as well as some constructs that cannot be proven safe.
+
+Use of these potentially "unsafe" constructs can threaten the integrity of the .NET runtime stack, enabling modification
+of key information on a stack frame, like the values of code and data addresses.
+
+The .NET code generator includes stack frame protection (aka Guard Stack or "GS") so that the integrity of the stack
+can be checked at key points during program execution&mdash;walking the stack for EH or GC, or returning from methods.
+
+## How GS Works
+
+GS is intended to detect buffer overruns from unsafe on-stack buffers that might corrupt vulnerable data on the stack.
+
+Unsafe buffers include:
+* memory regions allocated dynamically on the stack, via `stackalloc` in C# (aka `localloc`, in IL)
+* value classes marked as unsafe by language compilers, via `System.Runtime.CompilerServices.UnsafeValueTypeAttribute`.
+For instance, C# [fixed-sized buffers](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/unsafe-code#fixed-size-buffers).
```

Comment:
Should this cover that other types of buffers (InlineArrays, creating spans to stack memory, etc)?

There's been discussion a few times about expanding the scenarios we emit the stack cookies to cover these other cases as well, even though it isn't done today.

Reply:
Right, they don't but it's been suggested previously that we do it for them as well. So it might be worth calling it out and providing examples.

The cookie is cheap and the benefits for unsafe code is typically worth it. Ideally the JIT could then elide in cases we know are bounds checked/safe, but that is a separate consideration.

Reply:
This writeup is intended to document what we do today (in particular what exists in shipping code). 

We can discuss possible enhancements or extensions separately.